aesara/tensor/shape.py

```diff
@@ -426,13 +443,18 @@ def make_node(self, x, shape):
                 f"Input `x` is {x.ndim}-dimensional and will never match a shape of length {len(shape)}."
             )
 
-        if isinstance(x.type, TensorType) and all(isinstance(s, Number) for s in shape):
-            out_var = TensorType(x.type.dtype, shape)()
+        if isinstance(x.type, TensorType):
+            mixed_shape = tuple(
+                xts if xts is not None else s for xts, s in zip(x.type.shape, shape)
+            )
+            folded_shape = constant_fold_shape(mixed_shape)
+            folded_valid_shape = tuple(s if s != -1 else None for s in folded_shape)
+            out_var = TensorType(x.type.dtype, folded_valid_shape)()
```

Comment:
The comment above mentions one way in which the need for constant folding can be reduced, but, more importantly, we really shouldn't need constant folding in order to find the static/non-static shape values.

Instead, if we use the established `None` encoding, then we can use a direct `NoneConst.equals` comparison with other symbolic terms (like the elements of `shape`) and it will work without constant folding.